### PR TITLE
disconnects connection pools on close by default closes #495

### LINKED ISSUE #495: Close method does nothing by default and undocumented setting.
Hi everyone. The method close in DefaultClient does nothing unless the setting DJANGO_REDIS_CLOSE_CONNECTION is at True, in this case, it closes its connection pool as it can be seen here:
https://github.com/jazzband/django-redis/blob/a7ca6e40fd487016cac819cc00ab19286bf3f063/django_redis/client/default.py#L569
I wonder why the closing of the connection pool is not the default behavior and why DJANGO_REDIS_CLOSE_CONNECTION is not documented.
The current behavior causes a great number of unneeded open connections to redis.
I suggest to remove DJANGO_REDIS_CLOSE_CONNECTION and to close the connections or at least, to document this setting if it is to remain.

### PR DESCRIPTION
**Pull request by pinoatrome**
configurable by settings
tried to document

### CODE REVIEW

File: django_redis/client/default.py [hunk 1]
```diff
@@ -566,12 +566,12 @@ def make_pattern(self, pattern, version=None, prefix=None):
...
-            for i in range(len(self._clients)):
-                client = self._clients[i]
-                if client is not None:
-                    client.connection_pool.disconnect()
-                self._clients[i] = None
+        close_flag = self._options.get(
+            "CLOSE_CONNECTION", getattr(settings, "DJANGO_REDIS_CLOSE_CONNECTION", True)
+        )
+        if close_flag:
+            [client.connection_pool.disconnect() for client in self._clients if client]
```
Comment (WisdomPill) [inline django_redis/client/default.py:573]: This is bad for readability, the good old for is easier to read and equivalent

File: tests/test_backend.py [hunk 2]
```diff
@@ -1117,6 +1122,37 @@ def test_actual_expiry(self):
 
 
 class TestDefaultClient(unittest.TestCase):
+    def setUp(self):
+        self.client = caches[DEFAULT_CACHE_ALIAS].client
+
+    def tearDown(self):
+        try:
+            self.client.clear()
+        except exceptions.ConnectionInterrupted:
```
Comment (WisdomPill) [inline tests/test_backend.py:1131]: No exception should pass silently, why is this here?

File: tests/test_backend.py [hunk 3]
```diff
@@ -1117,6 +1122,37 @@ def test_actual_expiry(self):
...
+            assert mock.called
+
+    @override_settings(DJANGO_REDIS_CLOSE_CONNECTION=False)
+    def test_close_disconnect_settings(self):
+        self.client.get_client(write=True)
+        with patch("redis.connection.ConnectionPool.disconnect") as mock:
+            self.client.close()
+            assert not mock.called
+
+    def test_close_disconnect_client_options(self):
```
Comment (WisdomPill) [inline tests/test_backend.py:1149]: Why not change the options, via `override_settings`? it is easier to read for the people that come after you

File: tests/test_backend.py [hunk 4]
```diff
@@ -1117,6 +1122,37 @@ def test_actual_expiry(self):
...
+    def test_close_client_disconnect_default(self):
+        self.client.get_client(write=True)
+        with patch("redis.connection.ConnectionPool.disconnect") as mock:
+            self.client.close()
+            assert mock.called
+
+    @override_settings(DJANGO_REDIS_CLOSE_CONNECTION=False)
+    def test_close_disconnect_settings(self):
+        self.client.get_client(write=True)
+        with patch("redis.connection.ConnectionPool.disconnect") as mock:
```
Comment (WisdomPill) [inline tests/test_backend.py:1145]: I do not know if this works with other clients

File: README.rst [hunk 5]
```diff
@@ -745,6 +745,29 @@ setting ``REDIS_CLIENT_KWARGS``.
...
     }
 
+
+Closing Connections
+~~~~~~~~~~~~~~~~~~~
+
+django-redis closes all the connections to Redis server by default, irrespectively of its current usage.
+
+You can change this behaviour for all caches by setting ``DJANGO_REDIS_CLOSE_CONNECTION = False``
+in the django settings or at cache level by setting ``CLOSE_CONNECTION: False`` in the OPTIONS
```
Comment (WisdomPill) [inline README.rst:755]: It is still too early to set it as true by default

**Review by WisdomPill — CHANGES_REQUESTED** [review]
Maybe only docs to show that there is the flag are enough for now, I would wait to see what other people say about the issue.

**Comment by WisdomPill** [comment]
If tests do not pass then there will be no merge

**Comment by WisdomPill** [comment]
All the tests are failing for the same reason.

```python
2021-01-15T13:42:42.1186259Z py36-dj22-redislatest run-test: commands[0] | /home/runner/work/django-redis/django-redis/.tox/py36-dj22-redislatest/bin/python -b -Wa tests/runtests.py
2021-01-15T13:42:42.1188525Z [3100] /home/runner/work/django-redis/django-redis$ /home/runner/work/django-redis/django-redis/.tox/py36-dj22-redislatest/bin/python -b -Wa tests/runtests.py
2021-01-15T13:43:01.8693138Z ...............................................................................................F.........
2021-01-15T13:43:01.8694312Z ======================================================================
2021-01-15T13:43:01.8695393Z FAIL: test_close_disconnect_client_options (tests.test_backend.TestDefaultClient)
2021-01-15T13:43:01.8697247Z ----------------------------------------------------------------------
2021-01-15T13:43:01.8698174Z Traceback (most recent call last):
2021-01-15T13:43:01.8700519Z   File "/home/runner/work/django-redis/django-redis/tests/test_backend.py", line 1154, in test_close_disconnect_client_options
2021-01-15T13:43:01.8701817Z     assert not mock.called
2021-01-15T13:43:01.8702548Z AssertionError
2021-01-15T13:43:01.8704700Z 
2021-01-15T13:43:01.8705602Z ----------------------------------------------------------------------
2021-01-15T13:43:01.8706146Z Ran 105 tests in 19.172s
2021-01-15T13:43:01.8706416Z 
2021-01-15T13:43:01.8706776Z FAILED (failures=1)
```

File: django_redis/client/default.py (hunk 1 above)
Comment (pinoatrome) [reply django_redis/client/default.py:573]: Agree, compound statements can be really bad for readability (just like most one-liners), still IMO list comprehension  is a great tool much appreciated for  brevity and  expressiveness when coded  in its simplest form  -> [do <smt> for <each> in iterable if <condition>] <- love it 🥇 
(here I'm trying to style code in the famous "pythonic way")
I will change it following your request anyway as I am not the mantainer.

PS
what about changing client.connection_pool.disconnect() with something like self.disconnect(idx) that follows the connect connect(idx) approach of delegating  to the connection factory connect/disconnect actions based on the configured client class? will propose an easy implementation based on this approach (there is always the opportunity to hear from your feedback :)

File: tests/test_backend.py (hunk 2 above)
Comment (pinoatrome) [reply tests/test_backend.py:1131]: hem...

File: tests/test_backend.py (hunk 4 above)
Comment (pinoatrome) [reply tests/test_backend.py:1145]: Correct and agree: solved with the definition of disconnect method on connection_factory (if that implementation makes sense)

File: README.rst (hunk 5 above)
Comment (pinoatrome) [reply README.rst:755]: done

File: tests/test_backend.py (hunk 3 above)
Comment (pinoatrome) [reply tests/test_backend.py:1149]: done

Comment (pinoatrome) [reply tests/test_backend.py:1161]: done

just for the sake of sharing: setting_changed signals is clearing the test fixture: 
needs to populate again the client connection(s).
```
    def test_close_disconnect_settings_cache(self):
        settings.CACHES[DEFAULT_CACHE_ALIAS]["OPTIONS"]["CLOSE_CONNECTION"] = True
        with override_settings(CACHES=settings.CACHES):
            # enabling override_settings context emits 'setting_changed' signal
            # (re-set the value to populate again client connections)
            self.cache.set("test_close_disconnect_settings_cache", 0)
```

**Review by WisdomPill — CHANGES_REQUESTED** [review]

File: django_redis/client/default.py (hunk 1 above)
Comment (WisdomPill) [reply django_redis/client/default.py]: this is not pythonic in my honest opinion, it would be if you would have used it as an array after, otherwise it is just writing less code and making it difficult to read, I invite you to read [the zen of python](https://www.python.org/dev/peps/pep-0020/).
In this regard, the first seven points describe the reason why I think the is not pythonic.

File: tests/test_backend.py (hunk 4 above)
Comment (WisdomPill) [reply tests/test_backend.py]: makes sense, nice catch. Because also the connect is delegated to the connection factory

File: django_redis/client/default.py (hunk 1 above)
Comment (pinoatrome) [reply django_redis/client/default.py]: done

File: README.rst [hunk 6]
```diff
@@ -749,13 +749,13 @@ setting ``REDIS_CLIENT_KWARGS``.
...
+django-redis doesn't close all the connections to Redis server on close().
 
-You can change this default behaviour for all caches by setting ``DJANGO_REDIS_CLOSE_CONNECTION = True``
-in the django settings or at cache level by setting ``CLOSE_CONNECTION: True`` in the ``OPTIONS``
+You can change this default behaviour for all caches by the ``DJANGO_REDIS_CLOSE_CONNECTION = True``
+in the django settings (globally) or (at cache level) by setting ``CLOSE_CONNECTION: True`` in the ``OPTIONS``
 for each configured cache.
 
-Setting True as a value will instruct the DefaultClient to close all the connections, irrespectively of its current usage.
+Setting True as a value will instruct the DefaultClient to close all the connections on close(), irrespectively of its current usage.
```
[the remark this answers is not in the document]
Comment (pinoatrome) [reply README.rst:758]: done
```
Setting True as a value will instruct the django-redis to close all the connections (since v. 4.12.2), irrespectively of its current usage.

```

File: tests/test_backend.py [hunk 7]
```diff
@@ -1121,41 +1121,51 @@ def test_actual_expiry(self):
         super().test_actual_expiry()
 
 
-class TestDefaultClient(unittest.TestCase):
+class TestClientClose(unittest.TestCase):
+    @property
+    def cache(self):
```
[the remark this answers is not in the document]
Comment (pinoatrome) [reply tests/test_backend.py:1126]: done

File: tests/test_backend.py [hunk 8]
```diff
@@ -1121,41 +1121,51 @@ def test_actual_expiry(self):
...
+        return caches[DEFAULT_CACHE_ALIAS]
+
+    @property
+    def client(self):
```
[the remark this answers is not in the document]
Comment (pinoatrome) [reply tests/test_backend.py:1130]: done

File: django_redis/client/sharded.py [hunk 9]
```diff
@@ -22,10 +21,10 @@ def __init__(self, *args, **kwargs):
         self._ring = HashRing(self._server)
         self._serverdict = self.connect()
 
-    def get_client(self, write=True):
+    def get_client(self, *args, **kwargs):
         raise NotImplementedError
 
-    def connect(self):
+    def connect(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (pinoatrome) [reply django_redis/client/sharded.py:27]: done, the change amends the previous implementation where the signature of connect does not match the signature of base method.

`def connect(self, index=0):`

**Review by WisdomPill — APPROVED** [review]
